lib/bolt/cli.rb

```diff
@@ -379,710 +309,563 @@ def validate(options)
-    rescue OptionParser::InvalidOption, OptionParser::AmbiguousOption => e
-      raise Bolt::CLIError, "Unknown argument '#{e.args.first}'"
-    end
+      unless %w[guide module project secret].include?(options[:subcommand]) ||
+             %w[convert new show].include?(options[:action]) ||
+             options[:plan_hierarchy]
+        target_opts   = options.slice(*TARGETING_OPTIONS)
+        target_string = "'--targets', '--rerun', or '--query'"
 
-    def puppetdb_client
-      plugins.puppetdb_client
+        if target_opts.length > 1
+          raise Bolt::CLIError, "Only one targeting option #{target_string} can be specified"
+        elsif target_opts.empty? && options[:subcommand] != 'plan'
+          raise Bolt::CLIError, "Command requires a targeting option: #{target_string}"
+        end
+      end
     end
 
-    def plugins
-      @plugins ||= Bolt::Plugin.setup(config, pal, analytics)
-    end
+    # Execute a Bolt command. The +options+ hash includes the subcommand and
+    # action to be run, as well as any additional arguments and options for the
+    # command.
+    #
+    # @param options [Hash] The CLI options.
+    #
+    def execute(options)
+      with_signal_handling do
+        with_error_handling do
+          # TODO: Separate from options hash and pass as own args.
+          command = options[:subcommand]
+          action  = options[:action]
+
+          #
+          # INITIALIZE CORE CLASSES
+          #
+
+          project = if ENV['BOLT_PROJECT']
+                      Bolt::Project.create_project(ENV['BOLT_PROJECT'], 'environment')
+                    elsif options[:project]
+                      dir = Pathname.new(options[:project])
+                      if (dir + Bolt::Project::BOLTDIR_NAME).directory?
+                        Bolt::Project.create_project(dir + Bolt::Project::BOLTDIR_NAME)
+                      else
+                        Bolt::Project.create_project(dir)
+                      end
+                    else
+                      Bolt::Project.find_boltdir(Dir.pwd)
+                    end
 
-    def query_puppetdb_nodes(query)
-      puppetdb_client.query_certnames(query)
-    end
+          config = Bolt::Config.from_project(project, options)
+
+          # TODO: Should this be an instance variable? Outputter is specific
+          #       to the CLI, so might make sense to have it be a property of
+          #       the class.
+          @outputter = Bolt::Outputter.for_format(
+            config.format,
+            config.color,
+            options[:verbose],
+            config.trace,
+            config.spinner
+          )
+
+          # TODO: Should this be an instance variable? Rerun is specific
+          #       to the CLI, so might make sense to have it be a property of
+          #       the class.
+          @rerun = Bolt::Rerun.new(config.rerunfile, config.save_rerun)
+
+          # TODO: Should this be an input to the executor? Might make sense
```

Comment:
Yeah I think having Analytics subscribe to the executor makes a lot of sense, and would mean that Skeletor wouldn't necessarily have to override all of these functions.